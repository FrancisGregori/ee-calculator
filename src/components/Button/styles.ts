import styled, { css } from 'styled-components'

export const Container = styled.button<{
  gridArea: string
  color?: string
  ref: any
}>`
  grid-area: ${({ gridArea }) => gridArea};
  display: flex;
  align-items: center;
  justify-content: center;
  background: #212121;
  box-shadow: 5px 5px 12px rgba(0, 0, 0, 0.25),
    -5px -5px 10px rgba(51, 51, 51, 0.25),
    inset -3px -3px 10px rgba(49, 49, 49, 0.25);
  height: 70px;
  width: 70px;
  border-radius: 35px;
  border: none;
  font-style: normal;
  font-weight: 400;
  font-size: 32px;
  color: ${({ color }) => color || '#848484'};
  outline: none !important;

  @media (max-width: 480px) {
    width: 48px;
    height: 48px;
    font-size: 24px;
  }

  ${({ gridArea }) =>
    (gridArea === 'plus' || gridArea === 'equal') &&
    css`
      height: 156px;
      @media (max-width: 480px) {
        height: 96px;
      }
    `}

  ${({ gridArea }) =>
    gridArea === 'zero' &&
    css`
      width: 156px;
      @media (max-width: 480px) {
        width: 104px;
      }
    `}
  
  &:focus,
  &:active {
    box-shadow: 4px 4px 4px rgba(27, 27, 27, 0.25),
      -2px -2px 5px rgba(37, 37, 37, 0.25),
      inset 4px 4px 10px rgba(10, 10, 10, 0.25),
      inset -6px -6px 10px rgba(49, 48, 48, 0.25);
  }
`
