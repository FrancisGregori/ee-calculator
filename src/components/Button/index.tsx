import React, { FC, memo, useCallback, useEffect, useRef } from 'react'

import { useCalculatorContext } from 'main/context/CalculatorProvider'
import { IKeypadButton } from 'main/types/keypadButton'

import { Container } from './styles'

const Button: FC<IKeypadButton> = ({ gridArea, value, color }) => {
  const buttonRef = useRef<HTMLButtonElement>()
  const { handleCalculation } = useCalculatorContext()

  const handleKeyDown = useCallback(
    ({ key, keyCode }: { key: string; keyCode: string | number }) => {
      if (key === value || (keyCode === 13 && value === '=')) {
        buttonRef.current?.focus()
        setTimeout(() => {
          buttonRef.current?.blur()
        }, 100)
      }
    },
    [value],
  )

  const handleOnClick = () => {
    handleCalculation(value)
    buttonRef.current?.blur()
  }

  useEffect(() => {
    window.addEventListener('keydown', handleKeyDown)

    return () => {
      window.removeEventListener('keydown', handleKeyDown)
    }
  }, [handleKeyDown])

  const handleLabel = () => {
    switch (value) {
      case '/':
        return <>&divide;</>
      case '*':
        return <>&times;</>
      case '-':
        return <>&minus;</>
      default:
        return value
    }
  }

  return (
    <Container
      color={color}
      ref={buttonRef}
      gridArea={gridArea}
      onClick={handleOnClick}
    >
      {handleLabel()}
    </Container>
  )
}

export default memo(Button)
