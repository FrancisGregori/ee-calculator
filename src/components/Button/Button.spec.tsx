import React from 'react'

import { render, screen } from '@testing-library/react'

import Button from './index'

describe('Button component', () => {
  it('should render correctly', async () => {
    const { asFragment } = render(<Button gridArea="clear" value="AC" />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should contain value AC', async () => {
    render(<Button gridArea="clear" value="AC" />)
    expect(screen.getByText('AC')).toBeTruthy()
  })

  it('should contain minus code', async () => {
    render(<Button gridArea="minus" value="-" />)
    expect(screen.getByText('−')).toBeTruthy()
  })

  it('should contain multiply code', async () => {
    render(<Button gridArea="multiply" value="*" />)
    expect(screen.getByText('×')).toBeTruthy()
  })

  it('should contain plus code', async () => {
    render(<Button gridArea="plus" value="+" />)
    expect(screen.getByText('+')).toBeTruthy()
  })

  it('should contain division code', async () => {
    render(<Button gridArea="divide" value="/" />)
    expect(screen.getByText('÷')).toBeTruthy()
  })
})
