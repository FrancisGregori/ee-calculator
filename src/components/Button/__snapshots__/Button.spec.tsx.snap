// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render correctly 1`] = `
<DocumentFragment>
  <button
    class="sc-bczRLJ gQUVyn"
  >
    AC
  </button>
</DocumentFragment>
`;
